browser/ui/brave_browser_actions.cc

```diff
@@ -0,0 +1,81 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/brave_browser_actions.h"
+
+#include "brave/components/ai_chat/core/common/buildflags/buildflags.h"
+#include "brave/components/playlist/common/buildflags/buildflags.h"
+#include "brave/components/vector_icons/vector_icons.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/views/side_panel/side_panel_action_callback.h"
+#include "chrome/browser/ui/views/side_panel/side_panel_entry_id.h"
+#include "chrome/browser/ui/views/side_panel/side_panel_entry_key.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/actions/actions.h"
+#include "ui/base/l10n/l10n_util.h"
+#include "ui/gfx/vector_icon_types.h"
+
+#if BUILDFLAG(ENABLE_AI_CHAT)
+#include "brave/components/ai_chat/core/browser/utils.h"
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
+
+#if BUILDFLAG(ENABLE_PLAYLIST)
+#include "brave/components/playlist/common/features.h"
+#endif  // BUILDFLAG(ENABLE_PLAYLIST)
+
+namespace {
+
+actions::ActionItem::ActionItemBuilder SidePanelAction(
+    SidePanelEntryId id,
+    int title_id,
+    int tooltip_id,
+    const gfx::VectorIcon& icon,
+    actions::ActionId action_id,
+    Browser* browser,
+    bool is_pinnable) {
+  return actions::ActionItem::Builder(CreateToggleSidePanelActionCallback(
+                                          SidePanelEntryKey(id), browser))
+      .SetActionId(action_id)
+      .SetText(l10n_util::GetStringUTF16(title_id))
+      .SetTooltipText(l10n_util::GetStringUTF16(tooltip_id))
+      .SetImage(ui::ImageModel::FromVectorIcon(icon, ui::kColorIcon))
+      .SetProperty(actions::kActionItemPinnableKey, is_pinnable);
+}
+
+}  // namespace
+
+BraveBrowserActions::BraveBrowserActions(Browser& browser)
+    : BrowserActions(browser) {
+  InitializeBraveBrowserActions();
+}
+
+BraveBrowserActions::~BraveBrowserActions() = default;
+
+void BraveBrowserActions::InitializeBraveBrowserActions() {
```

Comment:
again like other overrides where we want to replace the original implementation there's no reason to jump through hoops like this. Just Implement this in brave_actions.cc override and replace the original implementation

Reply:
cc: @simonhong

Reply:
This is for adding our two additional actions(playlist, chatUI) to upstream actions list.
I think you mean that replacing the file with the actions we only use.
Of course, upstream includes another actions we don't use but sometimes it's triggered in unexpected way.
Then, it could cause crash if actions list doesn't include action for it. Worrying about it.